editors/code/src/ctx.ts

```diff
@@ -15,29 +14,21 @@ export class Ctx {
     // FIXME: this actually needs syncronization of some kind (check how
     // vscode deals with `deactivate()` call when extension has some work scheduled
     // on the event loop to get a better picture of what we can do here)
-    client: lc.LanguageClient | null = null;
+    client: lc.LanguageClient;
     private extCtx: vscode.ExtensionContext;
 
-    constructor(extCtx: vscode.ExtensionContext) {
-        this.config = new Config(extCtx);
-        this.extCtx = extCtx;
-    }
-
-    async startServer() {
-        assert(this.client == null);
-
-        const client = await createClient(this.config);
-        if (!client) {
-            throw new Error(
-                "Rust Analyzer Language Server is not available. " +
-                "Please, ensure its [proper installation](https://github.com/rust-analyzer/rust-analyzer/tree/master/docs/user#vs-code)."
-            );
-        }
-
-        this.pushCleanup(client.start());
+    static async create(config: Config, extCtx: vscode.ExtensionContext, serverPath: string): Promise<Ctx> {
+        const client = await createClient(config, serverPath);
+        const res = new Ctx(config, extCtx, client);
+        res.pushCleanup(client.start());
         await client.onReady();
+        return res;
+    }
 
-        this.client = client;
+    private constructor(config: Config, extCtx: vscode.ExtensionContext, client: lc.LanguageClient) {
```

Comment:
@matklad , what about shortening the code?
```
private constructor(
    private readonly config: Config,
    private readonly extCtx: vscode.ExtensionContextx,
    private client: lc.LanguageClient
) {}
```

Reply:
wow, didn't realize that TypeScript can do that, thanks!